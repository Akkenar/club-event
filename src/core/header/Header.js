import React, { useContext, useEffect } from 'react';
import { Menu } from 'semantic-ui-react';
import { NavLink } from 'react-router-dom';
import getKey from '../intl/getKey';
import { setPageDescription } from '../../page.lib';
import configuration from '../../configuration';
import LanguageContext from '../intl/LanguageContext';
import Logo from '../logo/Logo';
import LanguageSwitcher from '../language-switcher/LanguageSwitcher';

import './Header.scss';

const Nav = props => <NavLink {...props} activeClassName={'active'} />;

const Header = () => {
  const { language, messages } = useContext(LanguageContext);

  // Not the best place, but works fine
  useEffect(() => {
    setPageDescription(getKey('description', messages));
  });

  const getUrlWithLanguage = url => {
    return `/${language}/${url}`;
  };

  return (
    <Menu inverted className="Header">
      <Nav to={getUrlWithLanguage('home')} color="white">
        <Logo className="Header__logo" width={60} height={45} inverted />
      </Nav>

      {configuration.registration.enabled ? (
        <Menu.Item as={Nav} name="register" to={getUrlWithLanguage('register')}>
          {getKey('header.register', messages)}
        </Menu.Item>
      ) : null}

      <LanguageSwitcher />
    </Menu>
  );
};

export default Header;
