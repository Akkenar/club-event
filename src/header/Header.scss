.Header {
  border-radius: 0 !important;
}

.Header__menu-container {
  margin: 0 0 0 auto;
}

.Header__logo {
  margin: 0 1em;
  padding-top: 0.2em;
  filter: invert(100%);
}
