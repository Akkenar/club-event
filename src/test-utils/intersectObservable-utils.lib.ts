import { mockGlobalProperty } from './test-utils.lib';

const allObservers: IntersectionObserverMock[] = [];

class IntersectionObserverMock {
  private readonly cb: any;
  public constructor(cb: any) {
    this.cb = cb;

    // To allow global triggering.
    allObservers.push(this);
  }

  public trigger() {
    this.cb([{ isIntersecting: true }]);
  }

  public observe() {
    // Nothing
  }

  public unobserve() {
    // Nothing.
  }

  public disconnect() {
    // Nothing.
  }
}

class IntersectionObserverEntryMock {
  public intersectionRatio() {
    // Nothing.
  }

  public isIntersecting() {
    // Nothing.
  }
}

export function mockInterestObservable(): () => void {
  // Reset all the mocks
  allObservers.slice(0);

  mockGlobalProperty(window)('IntersectionObserver')(IntersectionObserverMock);
  mockGlobalProperty(window)('IntersectionObserverEntry')(
    IntersectionObserverEntryMock,
  );

  // Trigger all.
  return () => {
    allObservers.forEach((observer: IntersectionObserverMock) => {
      observer.trigger();
    });
  };
}
