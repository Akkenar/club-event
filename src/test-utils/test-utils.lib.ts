import { fireEvent } from '@testing-library/react';

export function mockGrecaptcha() {
  (window as any).grecaptcha = {
    getResponse: () => 'response',
    render: jest.fn(),
    reset: jest.fn(),
  };
}

export function mockResponse(response: any) {
  const mock = typeof response === 'string' ? { result: response } : response;

  // Mock the fetch data as JSDom doesn't implement fetch.
  // @ts-ignore
  window.fetch = jest.fn(() =>
    Promise.resolve({
      text: () =>
        new Promise(resolve => {
          const mockResult = JSON.stringify(mock);
          setTimeout(() => {
            resolve(mockResult);
          }, 50);
        }),
    }),
  );
}

export function setValidInputData(
  validData: any,
  { getByTestId }: { getByTestId: any },
) {
  Object.keys(validData).forEach((key: string) => {
    fireEvent.change(getByTestId(key), {
      // @ts-ignore
      target: { value: validData[key] },
    });
  });
}

export const mockGlobalProperty = (globalObject: any) => (key: string) => (
  value: any,
) => {
  const original = globalObject[key];
  Object.defineProperty(globalObject, key, { value, writable: true });
  return () => mockGlobalProperty(globalObject)(key)(original);
};

export function forceLoadImage() {
  Object.defineProperty(Image.prototype, 'onload', {
    configurable: true,
    set(fn: () => void) {
      // Call the function directly to simulate the loading
      fn();
    },
  });
}
