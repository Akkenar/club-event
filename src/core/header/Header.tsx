import * as React from 'react';
import { useContext, useEffect } from 'react';
import { NavLink } from 'react-router-dom';
import { Menu } from 'semantic-ui-react';
import * as configuration from '../../configuration.json';
import getKey from '../intl/getKey';
import LanguageContext from '../intl/LanguageContext';
import LanguageSwitcher from '../language-switcher/LanguageSwitcher';
import Logo from '../logo/Logo';
import { hideSSRHeader, setPageDescription } from '../page.lib';

import './Header.scss';

const Nav = (props: any) => <NavLink {...props} activeClassName={'active'} />;

const Header = () => {
  const { language, messages } = useContext(LanguageContext);

  // Not the best place, but works fine
  useEffect(() => {
    setPageDescription(getKey('description', messages));
    // Only at this state we can hide the SSR. Doing it before will introduce a flicker.
    hideSSRHeader();
  });

  const getUrlWithLanguage = (url: string) => {
    return `/${language}/${url}`;
  };

  return (
    <Menu inverted={true} className="Header">
      <Nav to={getUrlWithLanguage('home')} color="white">
        <Logo className="Header__logo" width={60} height={45} inverted={true} />
      </Nav>

      {configuration.registration.enabled ? (
        <Menu.Item as={Nav} name="register" to={getUrlWithLanguage('register')}>
          {getKey('header.register', messages)}
        </Menu.Item>
      ) : null}

      <LanguageSwitcher />
    </Menu>
  );
};

export default Header;
