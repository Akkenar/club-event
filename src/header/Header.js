import React, { useContext, useEffect } from 'react';

import { Dropdown, Menu } from 'semantic-ui-react';
import { NavLink } from 'react-router-dom';
import getKey from '../intl/getKey';
import { setPageDescription } from '../page.lib';

import './Header.scss';
import logo from '../logo.svg';
import LanguageContext from '../intl/LanguageContext';

const Nav = props => <NavLink {...props} activeClassName={'active'} />;

const Header = () => {
  const { handleChangeLanguage, language, messages } = useContext(
    LanguageContext
  );

  // Not the best place, but works fine
  useEffect(() => {
    setPageDescription(getKey('description', messages));
  });

  return (
    <Menu>
      <Nav to={'/home/' + language}>
        <img
          className="Header__logo"
          alt={getKey('header.logo', messages)}
          src={logo}
          width={60}
          height={45}
        />
      </Nav>
      <Menu.Item as={Nav} name="home" to={'/home/' + language}>
        {getKey('header.information', messages)}
      </Menu.Item>

      <Menu.Item as={Nav} name="register" to={'/register/' + language}>
        {getKey('header.register', messages)}
      </Menu.Item>
      <Dropdown
        className="link item right Header__language-switcher"
        text={getKey('header.current.language', messages)}
      >
        <Dropdown.Menu>
          <Dropdown.Item
            title="FranÃ§ais"
            as={Nav}
            to="fr"
            onClick={() => handleChangeLanguage('fr')}
          >
            FR
          </Dropdown.Item>
          <Dropdown.Item
            title="Deutsch"
            as={Nav}
            to="de"
            onClick={() => handleChangeLanguage('de')}
          >
            DE
          </Dropdown.Item>
          <Dropdown.Item
            title="Italianno"
            as={Nav}
            to="it"
            onClick={() => handleChangeLanguage('it')}
          >
            IT
          </Dropdown.Item>
          <Dropdown.Item
            title="English"
            as={Nav}
            to="en"
            onClick={() => handleChangeLanguage('en')}
          >
            EN
          </Dropdown.Item>
        </Dropdown.Menu>
      </Dropdown>
    </Menu>
  );
};

export default Header;
