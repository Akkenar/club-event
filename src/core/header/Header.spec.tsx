import * as React from 'react';
import { HashRouter } from 'react-router-dom';
import { render, waitForElement } from '@testing-library/react';
import { forceLoadImage } from '../../test-utils/test-utils.lib';
import Header from './Header';

const HeaderWithRouter = () => (
  <HashRouter>
    <Header />
  </HashRouter>
);

forceLoadImage();

describe('Header', () => {
  it('should match snapshot without registration', async () => {
    const wrapper = render(<HeaderWithRouter />);
    await waitForElement(() => wrapper.getByTestId('logo'));
    expect(wrapper.baseElement).toMatchSnapshot();
  });
});
