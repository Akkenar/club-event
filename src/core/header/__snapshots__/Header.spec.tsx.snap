// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match snapshot without registration 1`] = `
<body>
  <div>
    <div
      class="ui inverted Header menu"
    >
      <a
        color="white"
        href="#/de/home"
      >
        <svg
          aria-label="logo.alt"
          class="Logo Logo--inverted  Header__logo"
          data-testid="lazyloaded-placeholder"
          height="45"
          viewBox="0 0 1024 1024"
          width="60"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d=""
          />
        </svg>
      </a>
      <a
        class="item"
        href="#/de/register"
      >
        header.register
      </a>
      <div
        aria-expanded="false"
        class="ui dropdown link item right LanguageSwitcher"
        data-testid="language-switcher"
        role="listbox"
        tabindex="0"
      >
        <div
          aria-live="polite"
          class="text"
          role="alert"
        >
          current.language
        </div>
        <i
          aria-hidden="true"
          class="dropdown icon"
        />
        <div
          class="menu transition"
        >
          <a
            class="item"
            href="#/fr/home"
            role="option"
            title="FranÃ§ais"
          >
            FR
          </a>
          <a
            class="item"
            href="#/de/home"
            role="option"
            title="Deutsch"
          >
            DE
          </a>
          <a
            class="item"
            href="#/it/home"
            role="option"
            title="Italianno"
          >
            IT
          </a>
          <a
            class="item"
            href="#/en/home"
            role="option"
            title="English"
          >
            EN
          </a>
        </div>
      </div>
    </div>
  </div>
</body>
`;
